interface/src/app/signature/enriched_signatures.js

```diff
@@ -0,0 +1,216 @@
+    if (!$stateParams.mlmodel) {
+      return;
+    }
+
+    self.modelInUrl = $stateParams.mlmodel;
+    self.statusMessage = 'Connecting to the server ...';
+    self.pValueCutoff = 0.05;
+    self.enrichedSignatures = [];
+
+    // Do nothing if no genes are specified in URL.
+    if (!$stateParams.genes || !$stateParams.genes.split(',').length) {
+      self.statusMessage = 'No genes are specified.';
+      self.enrichedSignatures = [];
+      return;
+    }
+
+    var genesInUrl = [];
+    $stateParams.genes.split(',').forEach(function(token) {
+      var id = parseInt(token);
+      if (!isNaN(id) && genesInUrl.indexOf(id) === -1) {
+        genesInUrl.push(id);
+      }
+    });
+
+    // Promise that gets all signatures of given mlmodel in URL:
+    var apiPromises = [];
+    apiPromises.push(Signature.get(
+      {mlmodel: self.modelInUrl, limit: 0},
+      function success(response) {
+        self.signatures = Object.create(null);
+        response.objects.forEach(function(element) {
+          self.signatures[element.id] = element.name;
+        });
+      },
+      function error(err) {
+        var message = errGen('Failed to get signatures: ', err);
+        $log.error(message);
+        self.statusMessage = message + '. Please try again later.';
+      }
+    ).$promise);
+
+    var participations;
+    // Promise that gets all participation records that are related to
+    // the genes in URL:
+    apiPromises.push(Participation.get(
+      {'related_genes': $stateParams.genes, 'limit': 0},
+      function success(response) {
+        participations = response.objects;
+      },
+      function error(err) {
+        var message = errGen(
+          'Failed to get signature-gene participations: ', err);
+        $log.error(message);
+        self.statusMessage = message + '. Please try again later.';
+      }
+    ).$promise);
+
+    // Helper function that calculates the enrichment for each signature
+    // that includes at least one gene in the url.
+    var calculateEnrichments = function(typeName, genesBySignatures) {
+      var m = genesInUrl.length,
+        N = self.geneNum,
+        pValueArray = [],
+        matchedGenesBySignature = [],
+        signatureIDs = Object.keys(genesBySignatures);
+      signatureIDs.forEach(function(sigID) {
+        var genes = genesBySignatures[sigID],
+          n = Object.keys(genes).length, k = 0,
+          matchedGenes = [], selectedGene, pValue;
+        for (var i = 0; i < genesInUrl.length; i++) {
+          selectedGene = genesInUrl[i];
+          if (genes[selectedGene]) {
+            k++;
+            matchedGenes.push(genes[selectedGene]);
+          }
+        }
+        matchedGenesBySignature.push(matchedGenes);
+
+        // Four parameters are needed in the formula:
+        //     hyperGeometricTest(k, m, n, N)
+        // k: the number of heavy genes related to sigID in genesInUrl;
+        // m: the number of genes in genesInUrl;
```

Comment:
Ah, I just thought of something for this parameter - do you think we need to check whether all of the genes in the url are found in any of the signatures for a specific model? This is because the `m` parameter is the number of "successes" in the population. Maybe each of the genes in the `genesInUrl` should only be considered a "success" if they are found somewhere in the population (meaning they have a participation tying them to one or more signatures for this model)?

Reply:
Hmmm, on second thought, I think this might be a moot point, since the `N` parameter below (which is the whole "population" of genes) takes into account ALL of the genes in the database for this organism (as we discussed), and the `genesInUrl` can only be drawn from that population of genes, right? So I think that as long as `N` is the total number of genes in the DB for that organism, this `m` parameter can just be the number of all genes in `genesInUrl`.